GH-3944: Merge feat private chain to dev

mpapierski opened the pull request:
Closes #3944 

This massive PR merges `feat-private-chain` branch into main code.

Commit:
Move system contracts logic into execution engine.

 Conflicts:
	execution_engine/src/core/runtime/auction_internal.rs
	execution_engine/src/system/auction.rs
	execution_engine/src/system/auction/detail.rs
	execution_engine/src/system/auction/providers.rs
	types/src/system/auction/unbonding_purse.rs

Commit:
Add API for enabling smart contracts.

Once a contract is disabled it can't be called unless it's enabled again
with this new API.

Commit:
Backport move TestRng into casper_types.

Commit:
Initial private chain support.

Commit:
Fix chainspecs after backporting.

Commit:
Revert accidentally removed contracts.

Commit:
Update private chainspec template.

Commit:
Fix assemblyscript build

casperlabs-bors-ng[bot] committed:
Merge #3106

3106: Initial support for private chain r=mpapierski a=mpapierski

This is backport of original #3029 on top of v1.4.6 with least amount of extra backported changes to make the original code from PR unchanged.


This PR brings preliminary support for private chains by introducing multiple chainspec configuration flags that change the behavior of a chain.

These settings are orthogonal to each other; any particular combination (and all other pre-existing settings) is valid for a given network, for a wide variety of possible permutations. 

- Administrator accounts: by specifying accounts in a `[[administrators]]` section of `accounts.toml` accounts with special privileges are created and later referred to as 'administrative accounts'
- `core.allow_auction_bids` disables add_bid, delegate entrypoints in auction contract if set to true
- `core.compute_rewards` rewards 0 tokens after each switch block. This means no new tokens are minted and validators aren't gaining more weight.
- `core.allow_unrestricted_transfers` if set to false, the normal user can't send tokens to other normal users; the normal user can transfer only to the administrator; either source or target has to be an admin.
- `core.refund_handling` defines how refunds are computed. 
  - `refund_handling = { type = "refund", refund_ratio = [1, 1] }` means 100% of `payment_amount - gas_spent` is refunded. 
  - `refund_handling = { type = "refund", refund_ratio = [0, 100 ] }` doesn't refund anything
  - `Refund` mode requires that the fraction specified is proper fraction (that is `refund_ratio <= 1`)
- `core.fee_handling` defines how gas spent is processed after deploy is executed.
  - `fee_handling = { type = "accumulate" }` creates special rewards purse in the mint system contract where fees are accumulated into
  - `fee_handling = { type = "pay_to_proposer" }` is a default setting for a public chain - rewards are transferred to block's proposer.
  - `fee_handling = { type = "burn" }` fees are burned.


Co-authored-by: Michał Papierski <michal@casperlabs.io>

Commit:
Merge tag 'v1.4.8' into feat-private-chain

 Conflicts:
	Cargo.lock
	execution_engine_testing/tests/src/test/regression/gov_74.rs
	execution_engine_testing/tests/src/wasm_utils.rs
	types/src/system/mod.rs

Commit:
Add support for CA signed client certs

Squashed commit of the following:

commit 04a0557bc2ca6076e7fea9d590cbc31126386a05
Author: Michał Papierski <michal@casperlabs.io>
Date:   Tue Aug 30 15:49:52 2022 +0200

    Remove ca_cert from the identity struct.

commit 8491fed80ae57c4ff4de6e99847f30b97220152b
Author: Michał Papierski <michal@casperlabs.io>
Date:   Tue Aug 30 15:14:02 2022 +0200

    Remove outdated TODO comment

commit 6060eed21200371613c46da0da15c117667fc3d7
Author: Michał Papierski <michal@casperlabs.io>
Date:   Tue Aug 30 15:12:17 2022 +0200

    Apply @marc-casperlabs comments

commit 2f04a174372efcc87c8049a0e566cb95c3607d6c
Author: Michał Papierski <michal@casperlabs.io>
Date:   Thu Aug 25 17:20:51 2022 +0200

    Remove unnecessary arg.

commit a1589039eec4311d90483836ac2b6e92ac0603f9
Author: Michał Papierski <michal@casperlabs.io>
Date:   Thu Aug 25 16:19:23 2022 +0200

    Simplify CA validation.

    This does not enforce ECC algorithm used, only basic expiration dates
    etc.

commit b98f0839c22c89de1facf2488973c9834a7866cb
Author: Michał Papierski <michal@casperlabs.io>
Date:   Thu Aug 25 14:00:25 2022 +0200

    Use #[source] instead of #[from]

casperlabs-bors-ng[bot] committed:
Merge #3310

3310: [private chain] Backport 1.4.8 fixes and support for CA signed client certificates  r=mpapierski a=mpapierski

This PR contains updates `feat-private-chain` for a new private chain release independent of 1.5.0:

- Merge of v1.4.8 security release
- Backport of #3291

After this PR is merged, it should be possible to release `private-1.4.8`.

Co-authored-by: Michał Papierski <michal@casperlabs.io>
Co-authored-by: Joe Sacher <321623+sacherjj@users.noreply.github.com>
Co-authored-by: Fraser Hutchison <fraser@casperlabs.io>
Co-authored-by: casperlabs-bors-ng[bot] <82463608+casperlabs-bors-ng[bot]@users.noreply.github.com>

Commit:
Merge branch 'feat-private-chain' into dev

 Conflicts:
	Cargo.lock
	execution_engine/src/core/engine_state/engine_config.rs
	execution_engine/src/core/engine_state/error.rs
	execution_engine/src/core/engine_state/genesis.rs
	execution_engine/src/core/engine_state/mod.rs
	execution_engine/src/core/engine_state/upgrade.rs
	execution_engine/src/core/execution/error.rs
	execution_engine/src/core/resolvers/v1_function_index.rs
	execution_engine/src/core/resolvers/v1_resolver.rs
	execution_engine/src/core/runtime/auction_internal.rs
	execution_engine/src/core/runtime/externals.rs
	execution_engine/src/core/runtime/handle_payment_internal.rs
	execution_engine/src/core/runtime/mint_internal.rs
	execution_engine/src/core/runtime/mod.rs
	execution_engine/src/core/runtime/scoped_instrumenter.rs
	execution_engine/src/core/runtime/standard_payment_internal.rs
	execution_engine/src/core/tracking_copy/ext.rs
	execution_engine/src/shared/host_function_costs.rs
	execution_engine/src/system/auction.rs
	execution_engine/src/system/auction/detail.rs
	execution_engine/src/system/auction/providers.rs
	execution_engine/src/system/handle_payment.rs
	execution_engine/src/system/handle_payment/internal.rs
	execution_engine/src/system/handle_payment/runtime_provider.rs
	execution_engine/src/system/mint.rs
	execution_engine/src/system/standard_payment.rs
	execution_engine/src/system/standard_payment/mint_provider.rs
	execution_engine_testing/test_support/src/auction.rs
	execution_engine_testing/test_support/src/lib.rs
	execution_engine_testing/test_support/src/wasm_test_builder.rs
	execution_engine_testing/tests/src/lmdb_fixture.rs
	execution_engine_testing/tests/src/test/regression/ee_966.rs
	execution_engine_testing/tests/src/test/regression/gh_2280.rs
	execution_engine_testing/tests/src/test/regression/gh_2605.rs
	execution_engine_testing/tests/src/test/regression/gov_74.rs
	execution_engine_testing/tests/src/test/regression/regression_20220303.rs
	execution_engine_testing/tests/src/test/storage_costs.rs
	execution_engine_testing/tests/src/test/system_contracts/auction/bids.rs
	execution_engine_testing/tests/src/test/system_contracts/genesis.rs
	execution_engine_testing/tests/src/test/system_contracts/upgrade.rs
	execution_engine_testing/tests/src/test/system_costs.rs
	execution_engine_testing/tests/src/test/wasmless_transfer.rs
	node/Cargo.toml
	node/src/components/block_proposer/deploy_sets.rs
	node/src/components/block_proposer/tests.rs
	node/src/components/block_validator/tests.rs
	node/src/components/chainspec_loader.rs
	node/src/components/consensus/config.rs
	node/src/components/consensus/era_supervisor.rs
	node/src/components/consensus/protocols/highway.rs
	node/src/components/consensus/protocols/highway/tests.rs
	node/src/components/consensus/validator_change.rs
	node/src/components/contract_runtime.rs
	node/src/components/deploy_acceptor.rs
	node/src/components/deploy_acceptor/tests.rs
	node/src/components/event_stream_server/sse_server.rs
	node/src/components/event_stream_server/tests.rs
	node/src/components/fetcher/tests.rs
	node/src/components/gossiper/gossip_table.rs
	node/src/components/gossiper/tests.rs
	node/src/components/linear_chain/pending_signatures.rs
	node/src/components/linear_chain/signature_cache.rs
	node/src/components/linear_chain/state.rs
	node/src/components/network/config.rs
	node/src/components/network/error.rs
	node/src/components/network/tests.rs
	node/src/components/small_network.rs
	node/src/components/small_network/message.rs
	node/src/components/small_network/tasks.rs
	node/src/components/storage/tests.rs
	node/src/crypto/asymmetric_key.rs
	node/src/crypto/asymmetric_key_ext.rs
	node/src/reactor/initializer.rs
	node/src/reactor/participating/tests.rs
	node/src/testing.rs
	node/src/testing/multi_stage_test_reactor/test_chain.rs
	node/src/tls.rs
	node/src/types.rs
	node/src/types/block.rs
	node/src/types/chainspec.rs
	node/src/types/chainspec/accounts_config.rs
	node/src/types/chainspec/accounts_config/account_config.rs
	node/src/types/chainspec/accounts_config/delegator_config.rs
	node/src/types/chainspec/accounts_config/validator_config.rs
	node/src/types/chainspec/activation_point.rs
	node/src/types/chainspec/core_config.rs
	node/src/types/chainspec/deploy_config.rs
	node/src/types/chainspec/global_state_update.rs
	node/src/types/chainspec/highway_config.rs
	node/src/types/chainspec/network_config.rs
	node/src/types/chainspec/protocol_config.rs
	node/src/types/deploy.rs
	node/src/types/node_id.rs
	node/src/utils/external.rs
	resources/local/chainspec.toml.in
	resources/production/chainspec.toml
	resources/test/valid/0_9_0/chainspec.toml
	resources/test/valid/0_9_0_unordered/chainspec.toml
	resources/test/valid/1_0_0/chainspec.toml
	smart_contracts/contract/src/ext_ffi.rs
	smart_contracts/contracts_as/test/authorized-keys/assembly/index.ts
	smart_contracts/contracts_as/test/set-action-thresholds/package.json
	types/Cargo.toml
	types/src/cl_type.rs
	types/src/contracts.rs
	types/src/lib.rs
	types/src/system.rs
	types/src/system/auction.rs
	types/src/system/error.rs
	types/src/timestamp.rs
	utils/global-state-update-gen/Cargo.toml
	utils/global-state-update-gen/src/auction_utils.rs
	utils/global-state-update-gen/src/main.rs
	utils/global-state-update-gen/src/validators.rs
	utils/nctl/sh/scenarios/chainspecs/bond_its.chainspec.toml.in
	utils/nctl/sh/scenarios/chainspecs/itst13.chainspec.toml.in
	utils/nctl/sh/scenarios/chainspecs/itst14.chainspec.toml.in
	utils/nctl/sh/scenarios/chainspecs/upgrade_scenario_1.chainspec.toml.in
	utils/nctl/sh/scenarios/chainspecs/upgrade_scenario_10.chainspec.toml.in
	utils/nctl/sh/scenarios/chainspecs/upgrade_scenario_3.chainspec.toml.in
	utils/nctl/sh/scenarios/chainspecs/upgrade_scenario_4.chainspec.toml.in
	utils/nctl/sh/scenarios/chainspecs/upgrade_scenario_5.chainspec.toml.in
	utils/nctl/sh/scenarios/chainspecs/upgrade_scenario_6.chainspec.toml.in
	utils/nctl/sh/scenarios/chainspecs/upgrade_scenario_7.chainspec.toml.in
	utils/nctl/sh/scenarios/chainspecs/upgrade_scenario_8.chainspec.toml.in
	utils/nctl/sh/scenarios/chainspecs/upgrade_scenario_9.chainspec.toml.in

Commit:
Fix deprecated warnings and tests.

Commit:
Merge branch 'dev' into merge-feat-private-chain-to-dev

Fraser999 commented on resources/local/chainspec.toml.in:
```diff
@@ -73,6 +73,29 @@ simultaneous_peer_requests = 5
 max_delegators_per_validator = 0
+# Allow auction entrypoints such as delegate, redelegate and add_bid to function properly.
+#
+# Changing this option makes sense only for private chains which dont need auctioning new validator slots.
```
```suggestion
# Setting this to false makes sense only on private chains which don't need to auction new validator slots.
```

mpapierski replied:
Addressed

Fraser999 commented on resources/local/chainspec.toml.in:
```diff
@@ -73,6 +73,29 @@ simultaneous_peer_requests = 5
+# Changing this option makes sense only for private chains which dont need auctioning new validator slots.
+allow_auction_bids = true
+# Allow peer to peer transfers between users. Setting this to false makes sense only on private chains.
+allow_unrestricted_transfers = true
```
This would be clearer as `restrict_transfers = false` IMO, but I know this is a merge PR and I guess we shouldn't make a change like this here.  Can we ticket this for future work though?

mpapierski replied:
Agreed if memory serves, it was named like that in one of the iterations. Ticketed

mpapierski replied:
#4164

Fraser999 commented on resources/local/chainspec.toml.in:
```diff
@@ -73,6 +73,29 @@ simultaneous_peer_requests = 5
+#                 administrator accounts
+#   'burn': fees are burned
+fee_handling = { type = 'pay_to_proposer' }
+administrators = []
```
This setting needs a comment.

mpapierski replied:
Addressed

Fraser999 reviewed (commented):
Initial brief review.  Will continue the review later.

Commit:
Update config files

Commit:
Address configuration changes

Commit:
Remove private chain example config.

These were used only for development purposes to test the private chain
settings in nctl environment.

Commit:
Add changelog entries.

Commit:
Merge branch 'dev' into merge-feat-private-chain-to-dev

 Conflicts:
	execution_engine_testing/tests/src/test/regression/gh_3710.rs

Fraser999 commented on ci/nightly-test.sh:
```diff
@@ -106,6 +106,13 @@ start_run_teardown "itst02.sh"
+start_run_teardown "itst06_private_chain.sh"
+start_run_teardown "itst07_private_chain.sh"
+start_run_teardown "itst11_private_chain.sh"
+# start_run_teardown "itst13_private_chain.sh"
+# start_run_teardown "itst14_private_chain.sh"
```
Can these be enabled?

mpapierski replied:
Ticketed fix to these scripts so they're safe to be enabled

Fraser999 commented on utils/nctl/sh/scenarios/itst01_private_chain.sh:
Wasn't this going to be changed to just a symlink?

mpapierski replied:
Fixed in 4f43df4725a40779a7054508d4959e115baf075

Commit:
Replace same tests with symlinks

This should make it easy for maintaining these tests.

Commit:
Remove tests from nightly tests

These tests are supposed to be run manually as for now

cspramit commented:
@darthsiroftardis or @alsrdn to review and close

alsrdn commented on execution_engine/src/core/engine_state/engine_config.rs:
```diff
@@ -128,4 +182,225 @@ impl EngineConfig {
+            RefundHandling::Refund { refund_ratio } | RefundHandling::Burn { refund_ratio } => {
+                debug_assert!(
+                    refund_ratio <= Ratio::one(),
+                    "refund ratio should be a proper fraction"
```
Why not panic or throw an error here? Also I would change this message to something like `refund ration should be in the range of [0, 1]`.

alsrdn commented on execution_engine/src/core/engine_state/genesis.rs:
```diff
@@ -45,9 +46,32 @@ use crate::{
+/// Default number of validator slots.
+pub const DEFAULT_VALIDATOR_SLOTS: u32 = 5;
+/// Default auction delay.
+pub const DEFAULT_AUCTION_DELAY: u64 = 3;
```
Any reason we want this as 3? We changed it to 1 a while back in most places.

mpapierski replied:
Good catch, changed

alsrdn commented on execution_engine/src/core/engine_state/genesis.rs:
```diff
@@ -45,9 +46,32 @@ use crate::{
+/// Default lock-in period of 90 days
+pub const DEFAULT_LOCKED_FUNDS_PERIOD_MILLIS: u64 = 90 * 24 * 60 * 60 * 1000;
+/// Default number of eras that need to pass to be able to withdraw unbonded funds.
+pub const DEFAULT_UNBONDING_DELAY: u64 = 14;
```
Same for unbonding_delay. We changed it to 7 a while back. Do we need it to be 14?

mpapierski replied:
Good catch, changed

alsrdn commented on execution_engine/src/core/engine_state/transfer.rs:
```diff
@@ -29,6 +29,23 @@ pub enum TransferTargetMode {
+/// A target mode indicates if a native transfer's arguments will resolve to an existing purse, or
+/// will have to create a new account first.
+#[derive(Copy, Clone, Debug, PartialEq)]
+pub(crate) enum NewTransferTargetMode {
```
Do we still need the old `TransferTargetMode`?

mpapierski replied:
It'd be a breaking change sadly, we should not break (and release new major version) until 2.0 cc @Fraser999

alsrdn commented on execution_engine/src/core/engine_state/transfer.rs:
```diff
@@ -220,11 +233,11 @@ impl TransferRuntimeArgsBuilder {
     ///
     /// Returns [`TransferTargetMode`] with a resolved variant.
-    fn resolve_transfer_target_mode<R>(
+    pub(super) fn resolve_transfer_target_mode<R>(
```
nit: comment needs to be adjusted: `/// Returns [`NewTransferTargetMode`] with a resolved variant.`

alsrdn commented on execution_engine_testing/tests/fixtures/release_1_4_5/global_state/data.lmdb:
would be useful to have a changelog for this file to understand what it contains.

alsrdn commented on node/CHANGELOG.md:
```diff
@@ -197,6 +197,11 @@ All notable changes to this project will be documented in this file.  The format
 
 ## 1.4.8
 
+### Added
```
Hmm, shouldn't this be at the top?

mpapierski replied:
Ouch, seems like automerge failed. Thankfully those are duplicated entries. Fixed

alsrdn commented on utils/global-state-update-gen/src/main.rs:
```diff
@@ -149,6 +151,39 @@ fn main() {
                         .help("The config file to be used for generating the update"),
                 ),
         )
+        .subcommand(
```
Should we maybe also make this part as the `generic_update`?

mpapierski replied:
I'm not sure it's worth the hassle tbh, admins are special but only if you want to run private chain under strict regulations. I'd prefer to not mix this with normal production code

alsrdn reviewed (commented):
Looks good to me, only have some questions and minor nits.

Commit:
Address review comments

mpapierski commented:
bors r+

mpapierski commented:
bors r+

Commit:
Ignore tests temprarily while waiting for #4193

mpapierski commented:
bors r+

Commit:
Fix typo

mpapierski commented:
bors r+
